R/testHelpers.R

```diff
@@ -0,0 +1,186 @@
+  # functions that are reused across tests
+  source(file.path(testDir, "functions.R"))
+  # testthat::test_check assumes we are in path/to/animint/tests
+  old <- getwd()
+  on.exit(setwd(old), add = TRUE)
+  setwd(dirname(testDir))
+  # avoid weird errors if this function is called via testhat::check()
+  # https://github.com/hadley/testthat/issues/144
+  Sys.setenv("R_TESTS" = "")
+  test_check("animint", filter = filter)
+}
+
+#' Kill child process(es) that may have been initiated in animint testing
+#'
+#' Read process IDs from a file and kill those process(es)
+#' 
+#' @seealso \link{tests_run}
+#' @export
+tests_exit <- function() {
+  res <- stop_binary()
+  e <- try(readLines(pid_file(), warn = FALSE), silent = TRUE)
+  if (!inherits(e, "try-error")) {
+    pids <- as.integer(e)
+    res <- c(res, tools::pskill(pids))
+  }
+  unlink(pid_file())
+  invisible(all(res))
+}
+
+#' Spawn a child R session that runs a 'blocking' command
+#' 
+#' Run a blocking command in a child R session (for example a file server or shiny app)
+#'
+#' @param directory path that the  server should map to.
+#' @param port port number to _attempt_ to run server on.
+#' @param code R code to execute in a child session
+#' @return port number of the successful attempt
+run_servr <- function(directory = ".", port = 4848,
+                      code = "servr::httd(dir=\"%s\", port=%d, browser=FALSE)") {
+  dir <- normalizePath(directory)
+  cmd <- sprintf(
+    # escape all the things!
+    paste0("'library(methods); cat(Sys.getpid(), file=\"%s\", sep=\"\\\\n\", append=TRUE);", code, "'"),
+    pid_file(), dir, port
+  )
+  output <- tempfile(fileext = "txt")
+  t <- suppressWarnings(system2("Rscript", c("-e", cmd), 
+                                stderr = output, stdout = output, wait = FALSE))
+  Sys.sleep(5)
+  o <- paste(readLines(output, warn = FALSE), "\n")
+  message(o)
+  success <- !any(grepl("Error", o, fixed = TRUE))
+  if (success) {
+    return(port)
+  } else {
+    # if not successful, try a new port
+    # do I need to kill the other R process that failed?
+    new_port <- port + seq.int(-10, 10)[sample.int(n = 21, size = 1)]
+    message("couldn't start a server on port ", port, "; trying ", 
+            new_port, " instead")
+    run_servr(directory = directory, port = new_port, code = code)
+  }
+}
+
+# --------------------------
+# Functions that are used in multiple places
+# --------------------------
+
+stop_binary <- function() {
+  if (exists("pJS")) {
+    pJS$stop()
+  } else if (exists("remDr")) {
+    # these methods are really queries to the server
+    # thus, if it is already shut down, we get some arcane error message
+    e <- try({
+      remDr$closeWindow()
+      remDr$closeServer()
+    })
+  }
+  TRUE
+}
+
+# file that will keep track of all processes were initiated during animint testing
+pid_file <- function() {
+  f <- file.path(find_test_path(), "pids.txt")
+  if (!file.exists(f)) file(f, "a")
+  f
+}
```

Comment:
One benefit of this approach (over storing them in an environment variable) is that if you quit the R session without doing `tests_exit()`, you can still get at those process IDs. Since it works, I'd prefer leave it as is for now. I might be changing some things in the future anyway (I've been giving some thought to writing my own package with selenium bindings)